File: javascript/data_structures/Queue.js
/**
 * Queue (FIFO)
 * Enqueue/Dequeue/Peek: O(1)
 *
 * Using an object with head/tail indices instead of an array
 * because shift() on arrays is O(n). This way both ends are O(1).
 */

class Queue {
  constructor() {
    this._storage = {};
    this._head = 0;
    this._tail = 0;
  }

  enqueue(item) {
    this._storage[this._tail] = item;
    this._tail++;
  }

  dequeue() {
    if (this.isEmpty()) {
      throw new Error('Cannot dequeue from empty queue');
    }
    const item = this._storage[this._head];
    delete this._storage[this._head];
    this._head++;
    return item;
  }

  peek() {
    if (this.isEmpty()) {
      throw new Error('Cannot peek empty queue');
    }
    return this._storage[this._head];
  }

  isEmpty() {
    return this._tail === this._head;
  }

  size() {
    return this._tail - this._head;
  }
}

module.exports = Queue;

File: python/data_structures/queue.py
"""
Queue (FIFO)
------------
Enqueue: O(1)
Dequeue: O(1)
Peek:    O(1)

Using deque internally because popping from the front of a regular
list is O(n) — it has to shift everything. deque is doubly linked
so both ends are O(1).
"""

from collections import deque


class Queue:
    def __init__(self):
        self._items = deque()

    def enqueue(self, item):
        self._items.append(item)

    def dequeue(self):
        if self.is_empty():
            raise IndexError("dequeue from empty queue")
        return self._items.popleft()

    def peek(self):
        if self.is_empty():
            raise IndexError("peek from empty queue")
        return self._items[0]

    def is_empty(self):
        return len(self._items) == 0

    def size(self):
        return len(self._items)

    def __repr__(self):
        return f"Queue({list(self._items)})"
